File: Increff/799.champagne-tower.java
/**
 * Algorithm:
 * Simulate row by row. Each glass keeps at most 1 cup; any overflow is split
 * equally to the two glasses below.
 *
 * Java data structures:
 * A double[] stores only the current row, reducing space from the full triangle
 * to O(query_row).
 *
 * Complexity:
 * Time O(query_row^2), space O(query_row).
 */
class Solution {
    public double champagneTower(int poured, int query_row, int query_glass) {
        double[] row = {poured};
        for (int r = 0; r < query_row; r++) {
            double[] next = new double[row.length + 1];
            for (int i = 0; i < row.length; i++) {
                double overflow = Math.max(0.0, row[i] - 1.0) / 2.0;
                next[i] += overflow;
                next[i + 1] += overflow;
            }
            row = next;
        }
        return Math.min(1.0, row[query_glass]);
    }
}

File: Increff/799.champagne-tower.go
package leetcode

// ChampagneTower799 simulates flow row by row. A glass keeps one cup and splits
// overflow equally to the two glasses below. Only the previous row is needed.
//
// Time: O(query_row^2). Space: O(query_row).
func ChampagneTower799(poured int, queryRow int, queryGlass int) float64 {
	row := []float64{float64(poured)}
	for r := 0; r < queryRow; r++ {
		next := make([]float64, len(row)+1)
		for i, amount := range row {
			overflow := (amount - 1.0) / 2.0
			if overflow < 0 {
				overflow = 0
			}
			next[i] += overflow
			next[i+1] += overflow
		}
		row = next
	}
	if row[queryGlass] > 1.0 {
		return 1.0
	}
	return row[queryGlass]
}
